server/src/main/scala/org/http4s/server/middleware/authentication/Authentication.scala

```diff
@@ -0,0 +1,37 @@
+package org.http4s
+package server
+package middleware
+package authentication
+
+import scalaz.concurrent.Task
+import org.http4s.headers.`WWW-Authenticate`
+
+/**
+ * Authentication instances are middleware that provide a
+ * {@link HttpService} with HTTP authentication.
+ */
+trait Authentication {
+  /**
+   * Check if req contains valid credentials. You may assume that
+   * getChallenge( req ) is called at most once for every req object
+   * (to allow for side-effects, e.g. the incrementation of a nonce
+   * counter in DigestAuthentication).
+   * @param req The request received from the client.
+   * @return If req contains valid credentials, None is returned.
+   *         Otherwise, Some(challenge) is returned. In this case,
+   *         challenge will be included in the HTTP 401 Unauthorized
+   *         response that is returned to the client.
+   *
+   */
+  def getChallenge(req: Request): Option[Challenge]
```

Comment:
If this returned a `Task[Option[Challenge]]`, that would compose asynchronously and avoid the `Await.result` calls that come from the actor.